### ISSUE TITLE
Can't use any ALT+... characters

### ISSUE THREAD

**Issue by McSlow** [paradoxxxzero/butterfly#112]
Perhaps i'm doing something completely wrong,
but currently i'm unable to type in characters like @, | etc which need ALT for composing.
Everything else works.

Butterfly is 2.0.3, tried it on Safari, Chrome on OSx10.11 and IE11, Chrome on Win7, no luck with any of them. 
No special configuration, the current docker butterfly image or example can reproduce it...

**Comment by paradoxxxzero** [comment]
What's your keyboard layout? On bépo/azerty ALT_Gr + ... works fine here.
You can try to break to this line https://github.com/paradoxxxzero/butterfly/blob/master/coffees/term.coffee#L1274 in the debugger to see what happens.

**Comment by McSlow** [comment]
Hi, just found the Part in the corresponding js.

It breaks when hitting @ and so on, so keys at least aren't catched away by the Browser.

Chrome's Debugger gives me the following values after hitting "@":
```
a:KeyboardEvent
altKey:true
bubbles:true
cancelBubble:false
cancelable:true
charCode:0
code:"KeyL"
ctrlKey:false
currentTarget:Window
defaultPrevented:false
detail:0
eventPhase:3
isTrusted:true
key:"@"
keyCode:76
keyIdentifier:"U+004C"
location:0
metaKey:false
path:Array[1]
repeat:false
returnValue:true
shiftKey:false
sourceCapabilities:InputDeviceCapabilities
srcElement:body.terminal.focus
target:body.terminal.focus
timeStamp:548564.385
type:"keyup"
view:Window
which:76
__proto__:KeyboardEvent
```

So on the first looks this seems to be quite ok. It just doesn't appear on the screen.

**Comment by McSlow** [comment]
I played a bit with the debugger. Interesting thing is that while debugging, no chraracter at all appears on the screen. It just steps over a lot of conditionals ( shift pressed? and stuff like this) and then returns to the blink routine...
BTW: I have a QWERTZ keyboard everywhere...

**Comment by McSlow** [comment]
Damned, thats it. qwerty works perfectly fine...

**Comment by paradoxxxzero** [comment]
When you are stepping, it's normal that nothing is immediatly printed. The
js just translates the keydown event to a character and sends it to the
shell. The shell sends back that char (or not) to the terminal and the
terminal displays it.

**Comment by McSlow** [comment]
yes, but with stepping, i of course meant hammering through the whole loop :)
I expected to see something at least when it has gone through the display functions. 

Some thoughts: 

- Unfortunately, @ and stuff like this is placed on Keys which need need the ALT modifier which is not the case on us qwerty boards. Otherwise french qwerty/azerty boards need alt, too for @ .

- From Safari's debugger i have seen that the ev.AltGraphKey returns false, afaik the right alt key should be altgraph, but might be that this is outdated and both alt's are equal today...

**Comment by paradoxxxzero** [comment]
Haha ok, I don't have a computer nearby right now but I will look into it
tomorrow and install à qwertz keyboard layout to see what happens!

**Comment by paradoxxxzero** [comment]
Ok bad news is @ in qwertz works here (AltGr + v).
Good news is that AltGr is not handled by the browser directly and butterfly gets the @ character from the keypress event and not the keydown, (sorry I told you to step in the wrong method). The keyPress handling function is here: https://github.com/paradoxxxzero/butterfly/blob/master/coffees/term.coffee#L1520 
It gives a key number of 124 which translates to @:

```javascript
String.fromCharCode(124)
"|"
```

So what happens for you?

**Comment by McSlow** [comment]
When I put the breakpoint there, it doesn't break when hitting @. 
It does on "normal" keys, and it does when I hit the bypass sequence before ( ALT-Z, then @ ).
charCode and the keycode is 64, code is "KeyL" ( which is correct, @ is sitting on the L on my keyboard  - a "Mac" Keyboard.) 

Same on a Windows 7 PC (keycode/charcode 64). Just the Code is "KeyQ" there. German "PC" Keyboards have @ sitting on the Q, little bit different.
That Windows PC has an Apple Keyboard, to make sure I also installed a standard "PC" Keyboard, but it's exactly the same.

I tried a Win10 Box, too, same game...

Btw.: Pipe (|) returns keyCode 124 here :) Doesn't work, too, but does not need to be prefixed with Alt-Z to hit the breakpoint.

**Comment by McSlow** [comment]
Just installed all the deps and launched the code from source and put it into --verbose --more mode..

This is what i get while hitting that @ key :)

```bash
[D 160812 22:11:24 terminal:282] WRIT<u'S\x1bl'
[D 160812 22:16:03 terminal:293] READ>'\x07'
```
It blurrs the screen for a short moment while hitting @

**Comment by paradoxxxzero** [comment]
Okay I think I may have missed something.
The  `\x1bl` is indeed an ALT + L which is sent as is (because your shell may want to catch this).
I have the same behaviour when I input ALT + L here,  ALT_Gr + L works as intended though and put a @.
Maybe you ALT_Gr key is interpreted as an ALT key.

What exactly is your QWERTZ layout ?
Are you on MacOs ?

**Comment by McSlow** [comment]
Hi, yes, I'm on OS X, Standard "German" QUERTZ Layout.
Currently pressing ALT-L (that's where @ is normally located)  triggers the 'ls' Shortcut, regardless if I'm using the left or right ALT key.
Debug is also the same for left and right:
```bash
[D 160906 19:39:31 terminal:282] WRIT<u'S\x1bl'
[...]
[D 160906 19:39:43 terminal:282] WRIT<u'S\x1bl'
```

best regards,
Björn

**Comment by paradoxxxzero** [comment]
Hum seems like an apple keyboard related issue. The thing is, I don't have one.
Apparently Alt is used as a compose key which conflicts with Alt used as shortcut key.

How does it work on other terminals, which key replaces the alt-shortcut key? Is it the `meta` js modifier?
